### PR TITLE
SAF Storage implementation

### LINKED ISSUE #120: Storage Access Framework Compatibility for API 29+
https://developer.android.com/preview/privacy/storage
https://stackoverflow.com/questions/58379543/cant-create-directory-in-android-10
https://stackoverflow.com/questions/57116335/environment-getexternalstoragedirectory-deprecated-in-api-level-29-java

### PR DESCRIPTION
**Pull request by chaneylc**
various warning fixes
some refactoring

# Description

App now starts by asking user to define a folder. After choosing a folder the migrator starts which is optional and allows user to choose an old folder to copy files from. This functionality can be repeated in the Preferences/General/Root Directory preference.

Upon migration, old photo/audio paths are converted to document file uris within the database after they are moved.

Fixes #120 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by trife** [comment]
- Change migrator button color to colorPrimary
- When selecting folder, if `.fieldbook` file exists, don't show Migrator
- Update strings to say `Field Book` instead of `Coordinate`
- `android:showAsAction="always"` works but `app:showAsAction="always"` doesn't and dumps everything into a menu button. Pixel 3 with Android 12
- Update `Root Directory` preference strings to match Prospector/Coordinate strings. Summary does not currently display selected folder.

**Comment by trife** [comment]
@BrapiCoordinatorSelby @timparsons This update is huge and has the potential to break a lot of things (but is necessary and has been a long time coming). I would greatly appreciate it if you had a few minutes to scan through some of the changes for anything that might have gone overlooked.

**Comment by BrapiCoordinatorSelby** [comment]
yup I can take a quick look now, and I will have more time next week if it can wait that long?

**Comment by trife** [comment]
We can wait until next week. I'm just trying to get it merged into develop before the hackathon.

**Review by BrapiCoordinatorSelby — COMMENTED** [review]
Looks good overall 👍 I didn't run any code for the review, so I hope all the relevant features have been tested.

File: app/src/main/java/com/fieldbook/tracker/activities/ConfigActivity.java [hunk 1]
```diff
@@ -165,21 +151,27 @@ public void onCreate(Bundle savedInstanceState) {
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/activities/ConfigActivity.java:150]: Yes, I mainly wanted to tackle the raw string usages in preference keys. I added a more exhaustive commit for these usages, but there are still many constants that can be replaced.

File: app/src/main/java/com/fieldbook/tracker/activities/ConfigActivity.java [hunk 2]
```diff
@@ -974,69 +932,96 @@ protected Integer doInBackground(Integer... params) {
             //if the export did not fail, create a zip file with the exported files
             if (!fail) {
 
+                String zipFileName = exportFileString + ".zip";
+
                 //if bundle is checked, zip the files with the media dir for the given study
                 if (bundleChecked) {
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/activities/ConfigActivity.java:937]: This was changed in another issue already, I'll make a note and revisit this after they are merged.

File: app/src/main/java/com/fieldbook/tracker/activities/FileExploreActivity.java [hunk 3]
```diff
@@ -1,154 +1,156 @@
...
+    private final Comparator<DocumentFile> mComparator = (f1, f2) -> {
+        if (f1.isDirectory() && !f2.isDirectory()) {
+            // Directory before non-directory
+            return -1;
+        } else if (!f1.isDirectory() && f2.isDirectory()) {
+            // Non-directory after directory
+            return 1;
+        } else {
+            // Alphabetic order otherwise
+            return f1.toString().compareToIgnoreCase(f2.toString());
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/activities/FileExploreActivity.java]: Good call, I added a commit to fix this.

File: app/src/main/java/com/fieldbook/tracker/activities/FileExploreActivity.java [hunk 4]
```diff
@@ -158,73 +160,87 @@ else if (chosenFile.equalsIgnoreCase("up") && !sel.exists()) {
...
 
-                    return !(sel.toString().contains(".fieldbook") || sel.toString().contains("severity.txt") || sel.toString().contains("sharedpref.xml"));
+                    if (excludedExtensions.contains(checkExtension(name))) continue;
 
+                    if (includedExtensions.contains(checkExtension(name)) || file.isDirectory()) {
+                        accepted.add(file);
+                        continue;
+                    }
+
+                    if (!(name.contains(".fieldbook") || name.contains("severity.txt") || name.contains("sharedpref.xml"))) {
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/activities/FileExploreActivity.java:198]: I think the point is to never include .fieldbook or the other asset files, so they would never be in the included extensions. This is just original FB code that I didn't want to change, I went ahead and made an issue to clean this up at another time.

File: app/src/main/java/com/fieldbook/tracker/activities/TraitEditorActivity.java [hunk 5]
```diff
@@ -908,9 +893,9 @@ public void onBackPressed() {
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/activities/TraitEditorActivity.java:892]: True, we have plans to change the traits activity in a future project.

File: app/src/main/java/com/fieldbook/tracker/brapi/service/BrAPIServiceV1.java [hunk 6]
```diff
@@ -90,7 +88,7 @@ public BrAPIServiceV1(Context context) {
         this.programsApi = new ProgramsApi(apiClient);
         this.trialsApi = new TrialsApi(apiClient);
         this.traitsApi = new ObservationVariablesApi(apiClient);
-        this.phenotypesApi = new PhenotypesApi(apiClient);
+        PhenotypesApi phenotypesApi = new PhenotypesApi(apiClient);
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/brapi/service/BrAPIServiceV1.java]: Removed

File: app/src/main/java/com/fieldbook/tracker/preferences/DatabasePreferencesFragment.java [hunk 7]
```diff
@@ -142,41 +147,50 @@ protected void onPreExecute() {
         @Override
         protected Integer doInBackground(Integer... params) {
 
-            //first check if the file to import is just a .db file
+            if (mChosenFile != null && mChosenFile.getName() != null) {
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/preferences/DatabasePreferencesFragment.java]: Added commit that removed the global variable.

File: app/src/main/java/com/fieldbook/tracker/storage/StorageMigratorFragment.kt [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+                                }
+                            }
+
+                            file.updateUri(docFile)
+
+                        }
+
+                    } else file.updateUri(checkDoc)
+
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/storage/StorageMigratorFragment.kt:145]: Yes this is the intended behavior.

File: app/src/main/java/com/fieldbook/tracker/traits/PhotoTraitLayout.java [hunk 9]
```diff
@@ -447,11 +466,16 @@ public void onClick(View view) {
...
-                    takePicture();
-                } else
-                    Utils.makeToast(getContext(),getContext().getString(R.string.traits_create_photo_maximum));
+                DocumentFile photosDir = DocumentTreeUtil.Companion.getFieldMediaDirectory(getContext(), "photos");
+
+                if (photosDir != null) {
+                    // Do not take photos if limit is reached
+                    if (m == 0 || photosDir.length() < m) {
+                        takePicture();
+                    } else
```
[the remark this answers is not in the document]
Comment (chaneylc) [reply app/src/main/java/com/fieldbook/tracker/traits/PhotoTraitLayout.java:479]: Very true, it wasn't apparent at first what that implementation was trying to do, I added a commit that replaces the directory length with the number of photos saved for that plot.

**Comment by chaneylc** [comment]
We can ignore the failed brapi tests for now.

**Pull request merged**